test/all_reporters.js

```diff
@@ -0,0 +1,111 @@
+// THE SOFTWARE.
+
+import { assert } from 'chai';
+import CompositeReporter from '../src/reporters/composite_reporter';
+import InMemoryReporter from '../src/reporters/in_memory_reporter';
+import NoopReporter from '../src/reporters/noop_reporter';
+import RemoteReporter from '../src/reporters/remote_reporter';
+import UDPSender from '../src/reporters/udp_sender';
+import MockLogger from './lib/mock_logger';
+import LoggingReporter from '../src/reporters/logging_reporter';
+import RSVP from 'rsvp';
+
+describe('All Reporters should', () => {
+    it ('have proper names', () => {
+        let loggingReporter = new LoggingReporter();
+        let inMemoryReporter = new InMemoryReporter();
+        inMemoryReporter.setProcess('service-name', []);
+        let noopReporter = new NoopReporter();
+        let remoteReporter = new RemoteReporter();
+        let compositeReporter = new CompositeReporter();
+
+        assert.equal(loggingReporter.name(), 'LoggingReporter');
+        assert.equal(inMemoryReporter.name(), 'InMemoryReporter');
+        assert.equal(noopReporter.name(), 'NoopReporter');
+        assert.equal(remoteReporter.name(), 'RemoteReporter');
+        assert.equal(compositeReporter.name(), 'CompositeReporter');
+    });
+
+    it ('flush, clear, and close are covered', (done) => {
+        let loggingReporter = new LoggingReporter();
+        let inMemoryReporter = new InMemoryReporter();
+        inMemoryReporter.setProcess('service-name', []);
+        let noopReporter = new NoopReporter();
+        let sender = new UDPSender();
+        sender.setProcess(inMemoryReporter._process);
+        let remoteReporter = new RemoteReporter(sender);
+        let reporters = [
+            loggingReporter,
+            inMemoryReporter,
+            noopReporter,
+            remoteReporter
+        ];
+
+        let reporter = new CompositeReporter(reporters);
+        let closeCalled = 0;
+        let flushCalled = 0;
+        let promises = [];
+        let executeCallbackLast = (callback, threshold) => {
+            let count = 0;
+            return () => {
+                count++;
+                if (count >= threshold) {
+                    if (callback) {
+                        promises.push(new RSVP.Promise((resolve, reject) => {
+                            resolve(callback);
+                        }));
+                    }
+                }
+            }
+        };
+
+        reporter.clear();
+        reporter.flush(executeCallbackLast(() => { flushCalled = 1; }, reporters.length));
+        reporter.close(executeCallbackLast(() => { closeCalled = 1; }, reporters.length));
+
+        sender = new UDPSender();
+        sender.setProcess(inMemoryReporter._process);
+        remoteReporter._sender = sender;
+
+        // covered without callbacks
+        reporter.flush();
+        reporter.close();
+
+        RSVP.all(promises).then((callbacks) => {
```

Comment:
this is still wrong. Your `promises` array can be empty here, because it's only populated in the callbacks. Simplify this test, testing one thing only.
```
let reporters = makeAllReporters();
let cReporter = new CompositeReporter(reporters);
let calls = 0;
cReporter.close(() => {
    calls++;
    assert.equal(calls, 1);
    done();
});
```

I am not 100% certain, but above test should catch the callback not being called, AND the callback being called more than once. You need to **verify** that by changing the code so that the test fails for both scenarios. I am not 100% certain about the latter, calling more than once, because if that happens the assertion would certainly fail, but the `done()` would've been called by then, so not sure how the framework would react to that (in fact it could happen after all tests have finished). 

Another option is to use http://sinonjs.org/. The validation in the test then becomes
```
var callback = sinon.spy();
cReporter.close(callback);
assert(callback.calledOnce);
```

Again, fudge the reporter code to prove that the test WILL fail if the callback is not called or called twice.